sycl/source/handler.cpp

```diff
@@ -93,122 +93,126 @@ event handler::finalize() {
     return MLastEvent;
   MIsFinalized = true;
 
-  std::shared_ptr<detail::kernel_bundle_impl> KernelBundleImpPtr = nullptr;
-  // If there were uses of set_specialization_constant build the kernel_bundle
-  KernelBundleImpPtr = getOrInsertHandlerKernelBundle(/*Insert=*/false);
-  if (KernelBundleImpPtr) {
-    // Make sure implicit non-interop kernel bundles have the kernel
-    if (!KernelBundleImpPtr->isInterop() &&
-        !MImpl->isStateExplicitKernelBundle()) {
-      kernel_id KernelID =
-          detail::ProgramManager::getInstance().getSYCLKernelID(MKernelName);
-      bool KernelInserted =
-          KernelBundleImpPtr->add_kernel(KernelID, MQueue->get_device());
-      // If kernel was not inserted and the bundle is in input mode we try
-      // building it and trying to find the kernel in executable mode
-      if (!KernelInserted &&
-          KernelBundleImpPtr->get_bundle_state() == bundle_state::input) {
-        auto KernelBundle =
+  const auto &type = getType();
+  if (type == detail::CG::Kernel) {
+    // If there were uses of set_specialization_constant build the kernel_bundle
+    std::shared_ptr<detail::kernel_bundle_impl> KernelBundleImpPtr =
+        getOrInsertHandlerKernelBundle(/*Insert=*/false);
+    if (KernelBundleImpPtr) {
+      // Make sure implicit non-interop kernel bundles have the kernel
+      if (!KernelBundleImpPtr->isInterop() &&
+          !MImpl->isStateExplicitKernelBundle()) {
+        kernel_id KernelID =
+            detail::ProgramManager::getInstance().getSYCLKernelID(MKernelName);
+        bool KernelInserted =
+            KernelBundleImpPtr->add_kernel(KernelID, MQueue->get_device());
+        // If kernel was not inserted and the bundle is in input mode we try
+        // building it and trying to find the kernel in executable mode
+        if (!KernelInserted &&
+            KernelBundleImpPtr->get_bundle_state() == bundle_state::input) {
+          auto KernelBundle =
+              detail::createSyclObjFromImpl<kernel_bundle<bundle_state::input>>(
+                  KernelBundleImpPtr);
+          kernel_bundle<bundle_state::executable> ExecKernelBundle =
+              build(KernelBundle);
+          KernelBundleImpPtr = detail::getSyclObjImpl(ExecKernelBundle);
+          setHandlerKernelBundle(KernelBundleImpPtr);
+          KernelInserted =
+              KernelBundleImpPtr->add_kernel(KernelID, MQueue->get_device());
+        }
+        // If the kernel was not found in executable mode we throw an exception
+        if (!KernelInserted)
+          throw sycl::exception(make_error_code(errc::runtime),
+                                "Failed to add kernel to kernel bundle.");
+      }
+
+      switch (KernelBundleImpPtr->get_bundle_state()) {
```

Comment:
getOrInsertHandlerKernelBundle(/*Insert=*/false);

may we got segfault here if insert is false?